File: waterinbox/android/app/src/main/java/com/example/waterinbox/sensor/SensorManager.kt
package com.example.waterinbox.sensor

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import com.example.waterinbox.math.BoxMath
import com.example.waterinbox.math.FusionConfig
import com.example.waterinbox.socket.SocketManager
import com.example.waterinbox.math.fuse_madgwick
import com.example.waterinbox.math.fuse_mahony3
import com.example.waterinbox.math.fuse_mahony6
import com.example.waterinbox.math.fuse_ekf
import kotlin.math.sqrt

data class SensorData(
    val gyro: FloatArray = floatArrayOf(0f, 0f, 0f),
    val accel: FloatArray = floatArrayOf(0f, 0f, 0f),
    val magnet: FloatArray = floatArrayOf(0f, 0f, 0f),

    val accelCorr: FloatArray = floatArrayOf(0f, 0f, 0f),
    val gyroCorr: FloatArray = floatArrayOf(0f, 0f, 0f),
    val magnetCorr: FloatArray = floatArrayOf(0f, 0f, 0f),

    val quaternion: FloatArray = floatArrayOf(0f, 0f, 0f, 1f),  // [qx,qy,qz,qw]
    val euler: FloatArray = floatArrayOf(0f, 0f, 0f),
    val axisAngle: FloatArray = floatArrayOf(0f, 0f, 0f, 0f),  // (axisX, axisY, axisZ, angleDeg)

    val accelRaw: FloatArray = floatArrayOf(0f, 0f, 0f),  // raw accelerometer (positive = anti-gravity)
    val gravity: FloatArray = floatArrayOf(0f, 0f, -1f),  // world gravity in box space, from quaternion
    val dt: Float = 0f,            // time step in seconds
    val waterPoly: List<FloatArray> = emptyList(),  // intersection polygon debug points
    val algoParams: FloatArray = floatArrayOf(0f, 0f, 0f),  // algorithm-specific params (displayed in UI)
    // World axis vectors rotated by quaternion into body space (for debug rendering)
    val worldAxisX: FloatArray = floatArrayOf(1f, 0f, 0f),  // [x,y,z] world X+ → body
    val worldAxisY: FloatArray = floatArrayOf(0f, 1f, 0f),  // [x,y,z] world Y+ → body
    val worldAxisZ: FloatArray = floatArrayOf(0f, 0f, 1f),  // [x,y,z] world Z+ → body
    val boatVertices: FloatArray = floatArrayOf(0f),  // 8 boat corner positions [x,y,z]×8, computed in BoxSpace.drawBoat
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as SensorData
        return gyro.contentEquals(other.gyro) && accel.contentEquals(other.accel) &&
                magnet.contentEquals(other.magnet) && quaternion.contentEquals(other.quaternion) &&
                euler.contentEquals(other.euler) && axisAngle.contentEquals(other.axisAngle) &&
                gravity.contentEquals(other.gravity) &&
                kotlin.math.abs(dt - other.dt) < 1e-6f
    }

    override fun hashCode(): Int {
        var result = gyro.contentHashCode()
        result = 31 * result + accel.contentHashCode()
        result = 31 * result + magnet.contentHashCode()
        result = 31 * result + quaternion.contentHashCode()
        result = 31 * result + euler.contentHashCode()
        result = 31 * result + axisAngle.contentHashCode()
        result = 31 * result + gravity.contentHashCode()
        result = 31 * result + dt.hashCode()
        return result
    }
}

class SensorProbe(context: Context) : SensorEventListener {

    private val sm = context.getSystemService(Context.SENSOR_SERVICE) as SensorManager
    private val gyro = sm.getDefaultSensor(Sensor.TYPE_GYROSCOPE)!!
    private val accel = sm.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)!!
    private val magnet = sm.getDefaultSensor(Sensor.TYPE_MAGNETIC_FIELD)

    // Box dimensions matching WaterRenderer (using screen pixels)
    private val metrics = context.resources.displayMetrics
    private val boxW = metrics.widthPixels.toFloat()
    private val boxH = metrics.heightPixels.toFloat()
    private val boxD = minOf(boxW, boxH) / 2f
    private val boxMath = BoxMath(boxW / 2f, boxH / 2f, boxD)

    private val _data = MutableStateFlow(SensorData())
    val data: StateFlow<SensorData> = _data.asStateFlow()

    // Quaternion state (updated by fusion)
    private var qW = 1f
    private var qX = 0f
    private var qY = 0f
    private var qZ = 0f

    private var lastGyrTimestamp = 0L
    private var lastDt = 0f

    // Sensor values (raw Android)
    private var gX = 0f; private var gY = 0f; private var gZ = 0f
    private var aX = 0f; private var aY = 0f; private var aZ = 0f
    private var mX = 0f; private var mY = 0f; private var mZ = 0f
    private var fixAX = 0f; private var fixAY = 0f; private var fixAZ = 0f  // negated accel = gravity dir
    private var fixGX = 0f; private var fixGY = 0f; private var fixGZ = 0f
    private var fixMX = 0f; private var fixMY = 0f; private var fixMZ = 0f

    fun start() {
        sm.registerListener(this, gyro, SensorManager.SENSOR_DELAY_GAME)
        sm.registerListener(this, accel, SensorManager.SENSOR_DELAY_GAME)
        sm.registerListener(this, magnet, SensorManager.SENSOR_DELAY_GAME)
    }

    fun stop() {
        sm.unregisterListener(this)
    }

    override fun onSensorChanged(event: SensorEvent) {
        when (event.sensor.type) {
            Sensor.TYPE_GYROSCOPE -> {
                val gyrTs = event.timestamp  // nanoseconds since boot (sensor time base)
                gX = event.values[0]; gY = event.values[1]; gZ = event.values[2]
                fixGX = gX; fixGY = gY; fixGZ = gZ

                if (lastGyrTimestamp != 0L) {
                    lastDt = ((gyrTs - lastGyrTimestamp) * 1e-9f).coerceIn(0.001f, 0.1f)
                    val result = when (FusionConfig.algorithm) {
                        "madgwick"  -> fuse_madgwick(fixGX, fixGY, fixGZ, fixAX, fixAY, fixAZ, qW, qX, qY, qZ, lastDt)
                        "mahony3"   -> fuse_mahony3(fixGX, fixGY, fixGZ, qW, qX, qY, qZ, lastDt)
                        "ekf"       -> fuse_ekf(fixGX, fixGY, fixGZ, fixAX, fixAY, fixAZ, qW, qX, qY, qZ, lastDt)
                        else        -> fuse_mahony6(fixGX, fixGY, fixGZ, fixAX, fixAY, fixAZ, qW, qX, qY, qZ, lastDt)
                    }
                    qX = result[0]; qY = result[1]; qZ = result[2]; qW = result[3]
                }
                lastGyrTimestamp = gyrTs

                emit(lastDt, gyrTs)
            }
            Sensor.TYPE_ACCELEROMETER -> {
                aX = event.values[0]; aY = event.values[1]; aZ = event.values[2]
                fixAX = -aX; fixAY = -aY; fixAZ = -aZ
            }
            Sensor.TYPE_MAGNETIC_FIELD -> {
                mX = event.values[0]; mY = event.values[1]; mZ = event.values[2]
                fixMX = mX; fixMY = mY; fixMZ = mZ
            }
        }
    }

    private fun emit(dt: Float, gyroTimestampNs: Long) {
        // ── 1. 欧拉角 (ZYX, C code convention) ──
        // e[0]=roll(around X), e[1]=pitch(around Y), e[2]=yaw(around Z)
        val test = qY * qW - qX * qZ  // different from YXZ!
        val roll: Float
        val pitch: Float
        val yaw: Float

        if (test > 0.499f) {
            roll = (Math.PI / 2f).toFloat()
            pitch = (-2f * kotlin.math.atan2(qX, qW)).toFloat()
            yaw = 0f
        } else if (test < -0.499f) {
            roll = (-Math.PI / 2f).toFloat()
            pitch = (2f * kotlin.math.atan2(qX, qW)).toFloat()
            yaw = 0f
        } else {
            val qx2 = qX * qX
            val qy2 = qY * qY
            val qz2 = qZ * qZ
            // roll: atan2(2(xw+yz), 1-2(x²+y²))
            roll = (kotlin.math.atan2(2f * (qX * qW + qY * qZ), 1f - 2f * (qx2 + qy2))).toFloat()
            // pitch: asin(2(yw-xz))
            pitch = (kotlin.math.asin(2f * test)).toFloat()
            // yaw: atan2(2(zw+xy), 1-2(y²+z²))
            yaw = (kotlin.math.atan2(2f * (qZ * qW + qX * qY), 1f - 2f * (qy2 + qz2))).toFloat()
        }

        // ── 2. 轴角 (C quaternion2axisandangle convention) ──
        // If qw < 0, negate vector part to get canonical form for display:
        // equivalent rotation with angle in [0°, 180°]
        val qx_d = if (qW < 0f) -qX else qX
        val qy_d = if (qW < 0f) -qY else qY
        val qz_d = if (qW < 0f) -qZ else qZ
        val qvLen = sqrt(qx_d * qx_d + qy_d * qy_d + qz_d * qz_d)
        val angleDeg: Float
        val axisX: Float
        val axisY: Float
        val axisZ: Float
        if (qvLen > 1e-6f) {
            angleDeg = (kotlin.math.acos(qW.toDouble().coerceIn(-1.0, 1.0)) * 2.0 * 180.0 / Math.PI).toFloat()
            val invLen = 1.0f / qvLen
            axisX = qx_d * invLen
            axisY = qy_d * invLen
            axisZ = qz_d * invLen
        } else {
            angleDeg = 0f
            axisX = 0f; axisY = 0f; axisZ = 1f
        }

        // ── 3. 重力向量 (四元数逆旋 from world to body) ──
        val gravityLocalX =  2f * (qW * qY - qZ * qX)
        val gravityLocalY = -2f * (qY * qZ + qW * qX)
        val gravityLocalZ = -1f + 2f * (qX * qX + qY * qY)

        // World axis vectors — exact port of C quaternion2bodyspaceworldaxis(qx,qy,qz,qw)
        // r=worldX+, f=worldY+, t=worldZ+ in body space
        val wxX =  1f - 2f*(qY*qY + qZ*qZ)
        val wxY =  2f * (qX*qY - qZ*qW)
        val wxZ =  2f * (qX*qZ + qY*qW)
        val wyX =  2f * (qX*qY + qZ*qW)
        val wyY =  1f - 2f*(qX*qX + qZ*qZ)
        val wyZ =  2f * (qY*qZ - qX*qW)
        val wzX =  2f * (qX*qZ - qY*qW)
        val wzY =  2f * (qY*qZ + qX*qW)
        val wzZ =  1f - 2f*(qX*qX + qY*qY)

        // ── 5. 推送 ──
        _data.value = SensorData(
            gyro = floatArrayOf(gX, gY, gZ),
            accel = floatArrayOf(aX, aY, aZ),
            magnet = floatArrayOf(mX, mY, mZ),
            gyroCorr = floatArrayOf(fixGX, fixGY, fixGZ),
            accelCorr = floatArrayOf(fixAX, fixAY, fixAZ),
            magnetCorr = floatArrayOf(fixMX, fixMY, fixMZ),
            quaternion = floatArrayOf(qX, qY, qZ, qW),
            euler = floatArrayOf(
                Math.toDegrees(roll.toDouble()).toFloat(),
                Math.toDegrees(pitch.toDouble()).toFloat(),
                Math.toDegrees(yaw.toDouble()).toFloat()
            ),
            axisAngle = floatArrayOf(axisX, axisY, axisZ, angleDeg),
            accelRaw = floatArrayOf(aX, aY, aZ),
            gravity = floatArrayOf(gravityLocalX, gravityLocalY, gravityLocalZ),
            dt = dt,
            waterPoly = boxMath.solveWaterPlane(floatArrayOf(gravityLocalX, gravityLocalY, gravityLocalZ)).polygon,
            worldAxisX = floatArrayOf(wxX, wxY, wxZ),
            worldAxisY = floatArrayOf(wyX, wyY, wyZ),
            worldAxisZ = floatArrayOf(wzX, wzY, wzZ),
            algoParams = when (FusionConfig.algorithm) {
                "madgwick" -> floatArrayOf(FusionConfig.madgwickBeta, 0f, 0f)
                "mahony6"  -> floatArrayOf(FusionConfig.mahonyKp, FusionConfig.mahonyKi, 0f)
                else       -> floatArrayOf(0f, 0f, 0f)
            },
        )
        // ── 6. Socket output ──
        SocketManager.onSensorData(
            qX, qY, qZ, qW,
            fixGX, fixGY, fixGZ,   // gyroC (corrected)
            fixAX, fixAY, fixAZ,    // accelC (corrected, negated)
            fixMX, fixMY, fixMZ,    // magC (corrected)
            gyroTimestampNs,
        )
    }

    override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {}
}

File: waterinbox/ios/waterinbox/sensor/SensorManager.swift
import Foundation
import CoreMotion
import Combine
import UIKit

class RenderState: ObservableObject {
    static let shared = RenderState()

    @Published var drawWorldAxes: Bool = true
    @Published var drawGravityArrow: Bool = true
    @Published var drawMagnetArrow: Bool = true
    @Published var drawBoat: Bool = true
    @Published var drawWaterSurface: Bool = true
    @Published var drawWaterBody: Bool = true

    private init() {}
}

class SensorManager: ObservableObject {
    static let shared = SensorManager()

    private let motionManager = CMMotionManager()
    private var lastGyrTimestamp: TimeInterval = 0
    private var lastDt: Float = 0

    // Quaternion state
//    @Published var qX: Float = 0
//    @Published var qY: Float = 0
//    @Published var qZ: Float = 0
//    @Published var qW: Float = 1

    @Published var data = SensorData()

    // Accelerometer reading (for arrow display)
    @Published var accelX: Float = 0
    @Published var accelY: Float = 0
    @Published var accelZ: Float = 0

    // Magnetometer reading (for arrow display)
    @Published var magX: Float = 0
    @Published var magY: Float = 0
    @Published var magZ: Float = 0

    private var qW_f: Float = 1
    private var qX_f: Float = 0
    private var qY_f: Float = 0
    private var qZ_f: Float = 0

    // Boat vertices updated by MetalViewController after each draw
    var latestBoatVertices: [Float] = Array(repeating: 0, count: 24)

    // BoxMath for water polygon computation
    private var boxMath: BoxMath!

    private init() {
        // Initialize boxMath with screen dimensions
        let screenScale = Float(UIScreen.main.scale)
        let screenBounds = UIScreen.main.bounds
        let hw = Float(screenBounds.width) * screenScale
        let hh = Float(screenBounds.height) * screenScale
        let boxD = min(hw, hh) / 2
        self.boxMath = BoxMath(hw: hw, hh: hh, d: boxD)
    }

    func start() {
        guard motionManager.isDeviceMotionAvailable else {
            print("Device motion not available")
            return
        }

        // Use device motion (fused sensor data)
        motionManager.deviceMotionUpdateInterval = 1.0 / 60.0
        motionManager.startDeviceMotionUpdates(to: .main) { [weak self] motion, error in
            guard let self = self, let motion = motion else { return }
            self.processMotion(motion)
        }
    }

    func stop() {
        motionManager.stopDeviceMotionUpdates()
    }

    private func processMotion(_ motion: CMDeviceMotion) {
        // Update sensor data
        let gx = Float(motion.rotationRate.x)
        let gy = Float(motion.rotationRate.y)
        let gz = Float(motion.rotationRate.z)

        let ax = Float(motion.userAcceleration.x)
        let ay = Float(motion.userAcceleration.y)
        let az = Float(motion.userAcceleration.z)

        let mx = Float(motion.magneticField.field.x)
        let my = Float(motion.magneticField.field.y)
        let mz = Float(motion.magneticField.field.z)

        // Gravity direction (normalized) for water surface calculation
        let grav = motion.gravity
        // userAcceleration + gravity = actual acceleration (in g, ~1.0 when stationary)
        let actualAccelX = Float(motion.userAcceleration.x) + Float(grav.x)
        let actualAccelY = Float(motion.userAcceleration.y) + Float(grav.y)
        let actualAccelZ = Float(motion.userAcceleration.z) + Float(grav.z)

        // Update config for arrow display (use actual acceleration, not normalized)
        FusionConfig.accelX = actualAccelX
        FusionConfig.accelY = actualAccelY
        FusionConfig.accelZ = actualAccelZ
        FusionConfig.magX = mx
        FusionConfig.magY = my
        FusionConfig.magZ = mz

        // Store for gravity arrow
        accelX = actualAccelX
        accelY = actualAccelY
        accelZ = actualAccelZ
    
        // Apply custom fusion algorithm using raw (unfused) gyro/accel + previous state
        let timestamp = motion.timestamp
        let dt = Float(timestamp - lastGyrTimestamp)
        lastGyrTimestamp = timestamp
        if dt <= 0 || dt > 1.0 { lastDt = 1.0 / 60.0 } else { lastDt = dt }

        // rawQ is always from iOS CMDeviceMotion (system-fused)
        let quat = motion.attitude.quaternion
        let rawQ: [Float] = [Float(quat.x), Float(quat.y), Float(quat.z), Float(quat.w)]

        // fusion
        var fusedQ: [Float] = rawQ
        if FusionConfig.algorithm != "ios" {
            let (qx, qy, qz, qw) = (qX_f, qY_f, qZ_f, qW_f)
            switch FusionConfig.algorithm {
            case "mahony3":
                let r = fuse_mahony3(gx: gx, gy: gy, gz: gz, qW: qw, qX: qx, qY: qy, qZ: qz, dt: lastDt)
                qX_f = r[0]; qY_f = r[1]; qZ_f = r[2]; qW_f = r[3]
                fusedQ = r
            case "mahony6":
                let r = fuse_mahony6(gx: gx, gy: gy, gz: gz, ax: -ax, ay: -ay, az: -az, qW: qw, qX: qx, qY: qy, qZ: qz, dt: lastDt)
                qX_f = r[0]; qY_f = r[1]; qZ_f = r[2]; qW_f = r[3]
                fusedQ = r
            case "madgwick":
                let r = fuse_madgwick(gx: gx, gy: gy, gz: gz, ax: -ax, ay: -ay, az: -az, qW: qw, qX: qx, qY: qy, qZ: qz, dt: lastDt)
                qX_f = r[0]; qY_f = r[1]; qZ_f = r[2]; qW_f = r[3]
                fusedQ = r
            case "ekf":
                let r = fuse_mahony3(gx: gx, gy: gy, gz: gz, qW: qw, qX: qx, qY: qy, qZ: qz, dt: lastDt)
                qX_f = r[0]; qY_f = r[1]; qZ_f = r[2]; qW_f = r[3]
                fusedQ = r
            default:
                break
            }
        }

        //fix yaw
        let fixedQ: [Float] = fixYaw(qx: fusedQ[0], qy: fusedQ[1], qz: fusedQ[2], qw: fusedQ[3])

        // Compute gravity in local box space from quaternion
        let fX = fixedQ[0]
        let fY = fixedQ[1]
        let fZ = fixedQ[2]
        let fW = fixedQ[3]

        let gravityLocalX =  2 * (fW * fY - fZ * fX)
        let gravityLocalY = -2 * (fY * fZ + fW * fX)
        let gravityLocalZ = -1 + 2 * (fX * fX + fY * fY)

        // Compute world axes
        let wxX =  1 - 2*(fY*fY + fZ*fZ)
        let wxY =  2 * (fX*fY - fZ*fW)
        let wxZ =  2 * (fX*fZ + fY*fW)
        let wyX =  2 * (fX*fY + fZ*fW)
        let wyY =  1 - 2*(fX*fX + fZ*fZ)
        let wyZ =  2 * (fY*fZ - fX*fW)
        let wzX =  2 * (fX*fZ - fY*fW)
        let wzY =  2 * (fY*fZ + fX*fW)
        let wzZ =  1 - 2*(fX*fX + fY*fY)

        // Compute Euler angles
        let test = fY * fW - fX * fZ
        var roll: Float = 0
        var pitch: Float = 0
        var yaw: Float = 0

        if test > 0.499 {
            roll = Float.pi / 2
            pitch = -2 * atan2(fX, fW)
            yaw = 0
        } else if test < -0.499 {
            roll = -Float.pi / 2
            pitch = 2 * atan2(fX, fW)
            yaw = 0
        } else {
            let qx2 = fX * fX
            let qy2 = fY * fY
            let qz2 = fZ * fZ
            roll = atan2(2 * (fX * fW + fY * fZ), 1 - 2 * (qx2 + qy2))
            pitch = asin(2 * test)
            yaw = atan2(2 * (fZ * fW + fX * fY), 1 - 2 * (qy2 + qz2))
        }

        // Axis-angle (canonical form)
        let qx_d = fW < 0 ? -fX : fX
        let qy_d = fW < 0 ? -fY : fY
        let qz_d = fW < 0 ? -fZ : fZ
        let qvLen = sqrt(qx_d * qx_d + qy_d * qy_d + qz_d * qz_d)

        var angleDeg: Float = 0
        var axisX: Float = 0, axisY: Float = 0, axisZ: Float = 1

        if qvLen > 1e-6 {
            angleDeg = Float(acos(Double(fW).clamped(to: -1...1))) * 2 * 180 / Float.pi
            let invLen = 1.0 / qvLen
            axisX = qx_d * invLen
            axisY = qy_d * invLen
            axisZ = qz_d * invLen
        }

        // Update SensorData
        // quatFixed = fused output (custom algo or iOS raw), quatFused = iOS raw
        let waterPoly = boxMath.solveWaterPlane(gravity: [gravityLocalX, gravityLocalY, gravityLocalZ]).polygon
        DispatchQueue.main.async {
            self.data = SensorData(
                gyro: [gx, gy, gz],
                accel: [ax, ay, az],
                magnet: [mx, my, mz],
                gyroCorr: [gx, gy, gz],
                accelCorr: [-ax, -ay, -az],
                magnetCorr: [mx, my, mz],
                quatFused: fusedQ,
                quatFixed: fixedQ,
                euler: [roll * 180 / Float.pi, pitch * 180 / Float.pi, yaw * 180 / Float.pi],
                axisAngle: [axisX, axisY, axisZ, angleDeg],
                accelRaw: [ax, ay, az],
                gravity: [gravityLocalX, gravityLocalY, gravityLocalZ],
                dt: self.lastDt,
                waterPoly: waterPoly,
                worldAxisX: [wxX, wxY, wxZ],
                worldAxisY: [wyX, wyY, wyZ],
                worldAxisZ: [wzX, wzY, wzZ],
                boatVertices: self.latestBoatVertices,
                algoParams: self.getAlgoParams()
            )
        }
    }

    func computeBoatVertices() -> [Float] {
        return latestBoatVertices
    }

    func getAlgoParams() -> [Float] {
        switch FusionConfig.algorithm {
        case "madgwick":
            return [FusionConfig.madgwickBeta, 0, 0]
        case "mahony6":
            return [FusionConfig.mahonyKp, FusionConfig.mahonyKi, 0]
        default:
            return [0, 0, 0]
        }
    }
}

struct SensorData {
    var gyro: [Float] = [0, 0, 0]
    var accel: [Float] = [0, 0, 0]
    var magnet: [Float] = [0, 0, 0]
    var gyroCorr: [Float] = [0, 0, 0]
    var accelCorr: [Float] = [0, 0, 0]
    var magnetCorr: [Float] = [0, 0, 0]
    var quatFused: [Float] = [0, 0, 0, 1]
    var quatFixed: [Float] = [0, 0, 0, 1]
    var euler: [Float] = [0, 0, 0]
    var axisAngle: [Float] = [0, 0, 0, 0]
    var accelRaw: [Float] = [0, 0, 0]
    var gravity: [Float] = [0, 0, -1]
    var dt: Float = 0
    var waterPoly: [[Float]] = []
    var worldAxisX: [Float] = [1, 0, 0]
    var worldAxisY: [Float] = [0, 1, 0]
    var worldAxisZ: [Float] = [0, 0, 1]
    var boatVertices: [Float] = Array(repeating: 0, count: 24)
    var algoParams: [Float] = [0, 0, 0]
}

extension Comparable {
    func clamped(to range: ClosedRange<Self>) -> Self {
        return min(max(self, range.lowerBound), range.upperBound)
    }
}
